pkg/providers/instance/suite_test.go

```diff
@@ -147,6 +200,95 @@ var _ = Describe("VMInstanceProvider", func() {
 		Entry("non-zonal", azureEnvNonZonal, cloudProviderNonZonal),
 	}
 
+	Context("metrics integration", func() {
+		BeforeEach(func() {
+			instancemetrics.VMCreateStartMetric.Reset()
+			instancemetrics.VMCreateFailureMetric.Reset()
+		})
+
+		It("records VM create start metric during successful launch", func() {
+			ExpectApplied(ctx, env.Client, nodeClaim, nodePool, nodeClass)
+			pod := coretest.UnschedulablePod(coretest.PodOptions{})
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, coreProvisioner, pod)
+			ExpectScheduled(ctx, env.Client, pod)
+
+			Expect(azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.CalledWithInput.Len()).To(BeNumerically(">=", 1))
+			createInput := azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.CalledWithInput.Pop()
+			labels := vmMetricLabelsFromCreateInput(createInput, nodePool.Name)
+
+			metric, ok, err := metrics.FindMetricWithLabelValues("karpenter_instance_vm_create_start_total", labels)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ok).To(BeTrue())
+			Expect(metric.GetCounter().GetValue()).To(BeNumerically("==", 1))
+
+			_, ok, err = metrics.FindMetricWithLabelValues("karpenter_instance_vm_create_failure_total", metrics.FailureMetricLabels(labels, "sync"))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ok).To(BeFalse())
+
+			_, ok, err = metrics.FindMetricWithLabelValues("karpenter_instance_vm_create_failure_total", metrics.FailureMetricLabels(labels, "async"))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ok).To(BeFalse())
+		})
+
+		It("records VM create sync failure metric when Azure returns an error", func() {
+			beginErr := &azcore.ResponseError{ErrorCode: "OperationNotAllowed"}
+			azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.BeginError.Set(beginErr)
+
+			ExpectApplied(ctx, env.Client, nodeClaim, nodePool, nodeClass)
+			pod := coretest.UnschedulablePod(coretest.PodOptions{})
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, coreProvisioner, pod)
+			ExpectNotScheduled(ctx, env.Client, pod)
+
+			Expect(azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.CalledWithInput.Len()).To(BeNumerically(">=", 1))
+			createInput := azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.CalledWithInput.Pop()
+			labels := vmMetricLabelsFromCreateInput(createInput, nodePool.Name)
+
+			metric, ok, err := metrics.FindMetricWithLabelValues("karpenter_instance_vm_create_start_total", labels)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ok).To(BeTrue())
+			Expect(metric.GetCounter().GetValue()).To(BeNumerically("==", 1))
+
+			syncFailureLabels := metrics.FailureMetricLabels(labels, "sync", map[string]string{metrics.ErrorCodeLabel: beginErr.Error()})
```

Comment:
Will need to fix this to expect the code is `OperationNotAllowed` not the whole err